pkg/controllers/node/termination/controller.go

```diff
@@ -104,17 +108,50 @@ func (c *Controller) Finalize(ctx context.Context, node *v1.Node) (reconcile.Res
 		}
 		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
 	}
-	// Be careful when removing this delete call in the Node termination flow
-	// This delete call is needed so that we ensure that we don't remove the node from the cluster
-	// until the full instance shutdown has taken place
-	if err := c.cloudProvider.Delete(ctx, nodeclaimutil.NewFromNode(node)); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
-		// We expect cloudProvider to emit a Retryable Error when the underlying instance is not terminated and if that
-		// happens, we want to re-enqueue reconciliation until we terminate the underlying instance before removing
-		// finalizer from the node.
-		if cloudprovider.IsRetryableError(err) {
+	nodeClaimList := &v1beta1.NodeClaimList{}
+	if err := c.kubeClient.List(ctx, nodeClaimList, client.MatchingFields{"status.providerID": node.Spec.ProviderID}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+	}
+	for i := range nodeClaimList.Items {
+		nodeClaim := &nodeClaimList.Items[i]
+		stored := nodeClaim.DeepCopy()
+		if nodeclaimutil.GetCondition(nodeClaim, v1beta1.Terminating).Status != v1.ConditionTrue {
+			// Be careful when removing this delete call in the Node termination flow
+			// This delete call is needed so that we ensure that we don't remove the node from the cluster
+			// until the full instance shutdown has taken place
+			if err := c.cloudProvider.Delete(ctx, nodeClaim); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
```

Comment:
Considering these are the exact same calls in the NodeCalim finalization and the node finalization flow, can we create a common function somewhere that just takes in the NodeClaim and does this all for us?